src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,903 @@
+  
+  MatType U, V;
+  U = L % (13 * 4 * (xyz.slice(0) % nz) - 13 * 0.197833) + 88 * maxi;
+  V = L % (13 * 9 * (xyz.slice(1) % nz) - 13 * 0.468331) + 134 * maxi;
+
+  CubeType OutImage(InImage.n_rows, InImage.n_cols, InImage.n_slices);
+  OutImage.slice(0) = L;
+  OutImage.slice(1) = U;
+  OutImage.slice(2) = V;
+  //OutImage = arma::join_slices(L,U);
+  //OutImage = arma::join_slices(OutImage, V);
+  return OutImage; 
+}
+
+template<typename MatType, typename CubeType>
+MatType StructuredForests<MatType, CubeType>::
+bilinearInterpolation(MatType const &src,
+                      size_t height, size_t width)
+{
+  MatType dst(height, width);
+  double const x_ratio = static_cast<double>((src.n_cols - 1)) / width;
+  double const y_ratio = static_cast<double>((src.n_rows - 1)) / height;
+  for(size_t row = 0; row != dst.n_rows; ++row)
+  {
+    size_t y = static_cast<size_t>(row * y_ratio);
+    double const y_diff = (row * y_ratio) - y; //distance of the nearest pixel(y axis)
+    double const y_diff_2 = 1 - y_diff;
+    for(size_t col = 0; col != dst.n_cols; ++col)
+    {
+      size_t x = static_cast<size_t>(col * x_ratio);
+      double const x_diff = (col * x_ratio) - x; //distance of the nearet pixel(x axis)
+      double const x_diff_2 = 1 - x_diff;
+      double const y2_cross_x2 = y_diff_2 * x_diff_2;
+      double const y2_cross_x = y_diff_2 * x_diff;
+      double const y_cross_x2 = y_diff * x_diff_2;
+      double const y_cross_x = y_diff * x_diff;
+      dst(row, col) = y2_cross_x2 * src(y, x) +
+              y2_cross_x * src(y, x + 1) +
+              y_cross_x2 * src(y + 1, x) +
+              y_cross_x * src(y + 1, x + 1);
+    }
+  }
+
+  return dst;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+sepFilter2D(CubeType& InImage, arma::vec& kernel, int radius)
+{
+  CubeType OutImage = this->CopyMakeBorder(InImage, radius, radius, radius, radius);
+
+  arma::vec row_res(1), col_res(1);
+  // reverse InImage and OutImage to avoid making an extra matrix.  
+  for(size_t k = 0; k < OutImage.n_slices; ++k)
+  {
+    for(size_t j = radius; j < OutImage.n_cols - radius; ++j)
+    {
+      for(size_t i = radius; i < OutImage.n_rows - radius; ++i)
+      {
+        row_res = OutImage.slice(k).row(i).subvec(j - radius, j + radius) * kernel;
+        col_res = OutImage.slice(k).col(i).subvec(i - radius, i + radius).t() * kernel;
+        // divide by 2: avg of row_res and col_res, divide by 3: avg over 3 locations.
+        InImage(i - radius, j - radius, k) = (row_res(0) + col_res(0)) / 2 / 3;
+      }
+    }
+  }
+
+  return InImage;
+}
+
+template<typename MatType, typename CubeType>
+CubeType StructuredForests<MatType, CubeType>::
+ConvTriangle(CubeType& InImage, int radius)
+{
+  if (radius == 0)
+  {
+    return InImage;
+  }
+  else if (radius <= 1)
+  {
+    double p = 12.0 / radius / (radius + 2) - 2;
+    arma::vec kernel = {1 , p, 1};
+    kernel = kernel / (p + 2);
```

Comment:
`kernel /= (p+2);`

Maybe can avoid copy, I am not sure about this part.Because armadillo may do a lot of jobs to avoid memory allocation in this casein this c